### PR TITLE
create parameter to determine whether roles are public or private

### PR DESCRIPTION
**Pull request by GaOrtiga**
### Description

ACS does not allow the operator to define the visibility of roles, meaning all of them are visible to all users, with the exception of Root Admin roles.

In order to address this situation, a new parameter `public_role` has been created in the following APIs: `createRole`, `updateRole` and `importRole`. This parameter adds a new property to the role, allowing it to be hidden from the users; so, when updating, importing or creating a new role it is possible for the operator to inform if it is public (visible to all users) or private (only visible to Root Admins and the creator of the role). Also, the behavior of the API `listRoles` has been adjusted to return the roles according to this new property.

The following are examples of the listings using a Root Admin account and a Domain Admin account.

<details><summary>Listing with a Root Admin account</summary>

```
(localcloud) 🐱 > list roles
{
  "count": 10,
  "role": [
    {
      "description": "Default root admin role",
      "id": "64b34ae1-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": false,
      "name": "Root Admin",
      "type": "Admin"
    },
    {
      "description": "Default resource admin role",
      "id": "64b38969-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Resource Admin",
      "type": "ResourceAdmin"
    },
    {
      "description": "Default domain admin role",
      "id": "64b48e57-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Domain Admin",
      "type": "DomainAdmin"
    },
    {
      "description": "Default user role",
      "id": "64b4ca77-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "User",
      "type": "User"
    },
    {
      "description": "Default read-only admin role",
      "id": "68ea06f5-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": false,
      "name": "Read-Only Admin - Default",
      "type": "Admin"
    },
    {
      "description": "Default read-only user role",
      "id": "68ea3701-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Read-Only User - Default",
      "type": "User"
    },
    {
      "description": "Default support admin role",
      "id": "68ea6ff9-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": false,
      "name": "Support Admin - Default",
      "type": "Admin"
    },
    {
      "description": "Default support user role",
      "id": "68eaab60-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": false,
      "name": "Support User - Default",
      "type": "User"
    },
    {
      "id": "2281d1f7-ba24-484e-bc1f-8519870dfc16",
      "isdefault": false,
      "ispublic": false,
      "name": "privaterole1",
      "type": "User"
    },
    {
      "id": "eebb8a56-6fce-4300-ade0-40a3e03a08a8",
      "isdefault": false,
      "ispublic": true,
      "name": "publicrole1",
      "type": "DomainAdmin"
    }
  ]
}
```
</details>

<details><summary>Listing with a Domain Admin account</summary>

```
(localcloud) 🐱 > list roles
{
  "count": 5,
  "role": [
    {
      "description": "Default resource admin role",
      "id": "64b38969-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Resource Admin",
      "type": "ResourceAdmin"
    },
    {
      "description": "Default domain admin role",
      "id": "64b48e57-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Domain Admin",
      "type": "DomainAdmin"
    },
    {
      "description": "Default user role",
      "id": "64b4ca77-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "User",
      "type": "User"
    },
    {
      "description": "Default read-only user role",
      "id": "68ea3701-26de-11ec-8dcf-5254005dcdac",
      "isdefault": true,
      "ispublic": true,
      "name": "Read-Only User - Default",
      "type": "User"
    },
    {
      "id": "eebb8a56-6fce-4300-ade0-40a3e03a08a8",
      "isdefault": false,
      "ispublic": true,
      "name": "publicrole1",
      "type": "DomainAdmin"
    }
  ]
}
```

</details>

Example of the other APIs:

<details><summary>createRole</summary>

```
(localcloud) 🐱 > create role roleid=64b34ae1-26de-11ec-8dcf-5254005dcdac name=testRole ispublic=false description="test role"
{
  "role": {
    "description": "test role",
    "id": "777e8dde-6670-42e2-8328-876d6445cc7c",
    "ispublic": false,
    "name": "testRole",
    "type": "Admin"
  }
}
```

</details>

<details><summary>updateRole</summary>

```
(localcloud) 🐱 > update role id=777e8dde-6670-42e2-8328-876d6445cc7c ispublic=true 
{
  "role": {
    "description": "test role",
    "id": "777e8dde-6670-42e2-8328-876d6445cc7c",
    "ispublic": true,
    "name": "testRole",
    "type": "Admin"
  }
}
```
</details>

<details><summary>importRole</summary>

```
(localcloud) 🐱 > import role name=importedRoleTest ispublic=false rules[0].rule=* rules[0].permission=allow type=User
{
  "role": {
    "id": "54045972-0272-4ab8-93c2-c299331897db",
    "ispublic": false,
    "name": "importedRoleTest",
    "type": "User"
  }
}
```
</details>

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?

1. I created two new roles using the `createRole` API. One with the parameter `public_role` set to `false` and one with it set to `true`. I verified that both roles were successfully created with the specified parameters.

2. I repeated the same steps from the first test, but this time using the `importRole` API instead. I verified that the roles were successfully imported with the specified parameters.

3. I updated the `public_role` parameter from two different roles using the `updateRole` API, the first one had the parameter as `true` and I updated it to `false` and the second had it as `false` and I updated it to `true`. Both were successfully updated.

4. I used the API `listRoles` using a Root Admin account and verified that every role was visible.

5. I used the API `listRoles` using a Domain Admin account and verified that it could not see private roles.

6. I created an account of the type user using a private role and verified that it could log in.

7. With this same account I created a new network, and a new VM, and verified that both worked normally.

8. I repeated tests 6 and 7 but this time with an account of the type Root Admin, and verified that it worked aswell.

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@GaOrtiga , can you look at the conflicts?

**Comment by GaOrtiga** [comment]
> @GaOrtiga , can you look at the conflicts?

Done.

**Review by JoaoJandre — APPROVED** [review]
CLGTM

**Comment by DaanHoogland** [comment]
Sorry @GaOrtiga , conflicts again. This happens a lot as we get close to release :(

**Comment by GaOrtiga** [comment]
> Sorry @GaOrtiga , conflicts again. This happens a lot as we get close to release :(

Yes, I resolved them, thanks!

File: api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ListRolesCmd.java [hunk 1]
```diff
@@ -110,7 +111,7 @@ private void setupResponse(final Pair<List<Role>, Integer> roles) {
     public void execute() {
         Pair<List<Role>, Integer> roles;
         if (getId() != null && getId() > 0L) {
-            roles = new Pair<List<Role>, Integer>(Collections.singletonList(roleService.findRole(getId())), 1);
+            roles = new Pair<List<Role>, Integer>(Collections.singletonList(roleService.findRole(getId(), true)), 1);
```
Comment (DaanHoogland) [inline api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ListRolesCmd.java:114]:
```suggestion
            roles = new Pair<>(Collections.singletonList(roleService.findRole(getId(), true)), 1);
```

Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ListRolesCmd.java:114]: no -1 by any means btw!

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@GaOrtiga do you think regression/integration tests as sensible for this change?

**Comment by GaOrtiga** [comment]
> @GaOrtiga do you think regression/integration tests as sensible for this change?

Do you mean for the PR as a whole or for this last change submitted?

If it is for the whole PR, I have run some regression tests and all of them worked out properly, however if you have concerns about any specific functionality, I can run some more.

If it is for this last change I don´t think any other testing is necessary, since the arguments deleted were really not being used.

**Comment by DaanHoogland** [comment]
No, I meant for the functionality in this PR in general. And I mean in an automated fashion so they can serve as regression tests in the future. I.e. automate the scenario you described under "How Has This Been Tested?".

**Comment by GaOrtiga** [comment]
@DaanHoogland Oh, I see.
Okay, I will look into it.

**Review by BryanMLima — APPROVED** [review]

**Comment by DaanHoogland** [comment]
test_affinity_groups_projects failing, see #7098

**Comment by DaanHoogland** [comment]
@rohityadavcloud @weizhouapache Can you guys give this a look? tnx

**Comment by DaanHoogland** [comment]
@GaOrtiga new conflicts. as we are closing in on release this will happen more often :(

**Comment by GaOrtiga** [comment]
> @GaOrtiga new conflicts. as we are closing in on release this will happen more often :(

Okay, I will look into it.
 I am currently working on the integration tests, took me way longer than expected, but hopefully will have them out by next week.

**Comment by DaanHoogland** [comment]
@GaOrtiga , i plan to cut the RC this weekend, will you be ready by today? (we need a day to run the tests as well.

**Comment by GaOrtiga** [comment]
@DaanHoogland I think its better if we push it to a future release.

File: test/integration/smoke/test_private_roles.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            self.domain_admin_apiclient
+        ]
+
+
+
+    def tearDown(self):
+        try:
+           cleanup_resources(self.apiclient, self.cleanup)
+        except Exception as e:
+            self.debug("Warning! Exception in tearDown: %s" % e)
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:161]:
```suggestion
        self.cleanup = []

        self.role_admin = Role.create(
            self.apiclient,
            self.testdata["roleadmin"]
        )
        self.cleanup.append(self.role_admin)

        self.role_domain_admin = Role.create(
            self.apiclient,
            self.testdata["roledomainadmin"]
        )
        self.cleanup.append(self.role_domain_admin)

        self.private_role = Role.create(
            self.apiclient,
            self.testdata["roleuser"]
        )
        self.cleanup.append(self.private_role)

        self.account_admin = Account.create(
            self.apiclient,
            self.testdata["accountadmin"],
            roleid=self.role_admin.id
        )
        self.cleanup.append(self.account_admin)

        self.account_domain_admin = Account.create(
            self.apiclient,
            self.testdata["accountdomainadmin"],
            roleid=self.role_domain_admin.id
        )
        self.cleanup.append(self.account_domain_admin)

        self.admin_apiclient = self.testClient.getUserApiClient(
            UserName=self.account_admin.name,
            DomainName='ROOT',
            type=1
        )

        self.domain_admin_apiclient = self.testClient.getUserApiClient(
            UserName=self.account_domain_admin.name,
            DomainName='ROOT',
            type=2
        )

    def tearDown(self):
        super(, self).tearDown()
```

File: test/integration/smoke/test_private_roles.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        self.testdata["roleuser"]["name"] += self.getRandomString()
+        self.testdata["publicrole"]["name"] += self.getRandomString()
+        private_role = Role.create(
+            self.apiclient,
+            self.testdata["roleuser"]
+        )
+        public_role = Role.create(
+            self.apiclient,
+            self.testdata["publicrole"]
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:216]:
```suggestion
        private_role = Role.create(
            self.apiclient,
            self.testdata["roleuser"]
        )
        self.cleanup.append(self.private_role)
        public_role = Role.create(
            self.apiclient,
            self.testdata["publicrole"]
        )
        self.cleanup.append(self.public_role)
```

File: test/integration/smoke/test_private_roles.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+            1. Create a public role
+            2. Check if its visibility is public
+            3. Update it to make it private
+            4. Verify if its visibility is private
+        """
+        self.testdata["publicrole"]["name"] += self.getRandomString()
+        role = Role.create(
+            self.apiclient,
+            self.testdata["publicrole"]
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:232]:
```suggestion
        role = Role.create(
            self.apiclient,
            self.testdata["publicrole"]
        )
        self.cleanup.append(role)
```

File: test/integration/smoke/test_private_roles.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        """
+            1. Import a public role
+            2. Import a private role
+            3. Verify their visibility
+        """
+        self.testdata["importrole"]["name"] += self.getRandomString()
+        imported_public_role = Role.importRole(
+            self.apiclient,
+            self.testdata["importrole"]
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:248]:
```suggestion
        imported_public_role = Role.importRole(
            self.apiclient,
            self.testdata["importrole"]
        )
        self.cleanup.append(imported_public_role)
```

File: test/integration/smoke/test_private_roles.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        self.testdata["importrole"]["name"] += self.getRandomString()
+        self.testdata["importrole"]["ispublic"] = False
+        imported_private_role = Role.importRole(
+            self.apiclient,
+            self.testdata["importrole"]
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:254]:
```suggestion
        imported_private_role = Role.importRole(
            self.apiclient,
            self.testdata["importrole"]
        )
        self.cleanup.append(imported_private_role)
```

File: test/integration/smoke/test_private_roles.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        """
+            1. Crate a User account with a private role
+            2. Login with the created account
+            3. Verify that the login was successful
+        """
+        account_private_role = Account.create(
+            self.apiclient,
+            self.testdata["accountroleuser"],
+            roleid=self.private_role.id
+        )
```
Comment (DaanHoogland) [inline test/integration/smoke/test_private_roles.py:270]:
```suggestion
        account_private_role = Account.create(
            self.apiclient,
            self.testdata["accountroleuser"],
            roleid=self.private_role.id
        )
        self.cleanup.append(account_private_role)
```

**Comment by GaOrtiga** [comment]
@DaanHoogland Can we proceed with this PR or is there any other concern regarding it?

**Comment by DaanHoogland** [comment]
> @DaanHoogland Can we proceed with this PR or is there any other concern regarding it?

no concerns

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
> @DaanHoogland this has been merged in `main` branch but the schema changes are in 4.18.0 to 4.18.1 upgrade path

thanks @shwstppr , I missed that. I can revert, but I see you added a commit on top of this. What do you propose we do?

**Comment by DaanHoogland** [comment]
Note how the smoke tests passed and didn't  warn me about my error. An upgrade test would have caught this :|

**Pull request merged**